crates/aptos-crypto/src/poseidon_bn254/mod.rs

```diff
@@ -0,0 +1,147 @@
+        7 => neptune_hash!(inputs, POSEIDON_7),
+        8 => neptune_hash!(inputs, POSEIDON_8),
+        9 => neptune_hash!(inputs, POSEIDON_9),
+        10 => neptune_hash!(inputs, POSEIDON_10),
+        11 => neptune_hash!(inputs, POSEIDON_11),
+        12 => neptune_hash!(inputs, POSEIDON_12),
+        13 => neptune_hash!(inputs, POSEIDON_13),
+        14 => neptune_hash!(inputs, POSEIDON_14),
+        15 => neptune_hash!(inputs, POSEIDON_15),
+        16 => neptune_hash!(inputs, POSEIDON_16),
+        _ => bail!(
+            "Poseidon-BN254 was called with {} inputs, more than the maximum 16 allowed inputs.",
+            inputs.len()
+        ),
+    };
+
+    Ok(result.into())
+}
+
+/// Given an string and `max_bytes`, it pads the byte array of the string with zeros up to size `max_bytes`,
+/// packs it to scalars, and returns the hash of the scalars.
+///
+/// This function calls `pad_and_pack_bytes_to_scalars_no_len` safely as strings will not contain the zero byte except to terminate.
+pub fn pad_and_hash_string(str: &str, max_bytes: usize) -> anyhow::Result<ark_bn254::Fr> {
+    pad_and_hash_bytes_with_len(str.as_bytes(), max_bytes)
+}
+
+/// Packs the bytes to a vector of scalars (see `pack_bytes_to_scalars`) and hashes the scalars via
+/// `hash_scalars`.
+///
+/// Note: The byte packing encodes the length of the bytes properly so as to avoid collisions when
+/// hashing, say, 0x00 versus 0x0000.
+///
+/// WARNING: We do not expose this function to avoid unnecessary bugs, since for SNARK circuits we
+/// always have to hash padded byte arrays. If necessary, an expert developer can indirectly call
+/// this via `pad_and_hash_bytes(bytes, bytes.len())`.
+#[allow(unused)]
+fn hash_bytes(bytes: &[u8]) -> anyhow::Result<ark_bn254::Fr> {
+    let scalars = pack_bytes_to_scalars(bytes)?;
+    hash_scalars(scalars)
+}
+
+/// Given `bytes`, if the length of `bytes` is less than `max_bytes`, pads `bytes` with zeros to length `max_bytes`.
+/// Then it packs padded `bytes` and returns the hash of the scalars via `hash_scalars`.
+///
+/// This is used when we know that `bytes` will not terminate in 0's and may skip encoding the length, for
+/// example ASCII strings. Otherwise unexpected collisions can occur.
+///
+/// Due to risk of collisions due to improper use by the caller, it is not exposed.
+#[allow(unused)]
+fn pad_and_hash_bytes_no_len(bytes: &[u8], max_bytes: usize) -> anyhow::Result<ark_bn254::Fr> {
+    let scalars = pad_and_pack_bytes_to_scalars_no_len(bytes, max_bytes)?;
+    hash_scalars(scalars)
+}
+
+/// Given `bytes`, if the length of `bytes` is less than `max_bytes`, pads `bytes` with zeros to length `max_bytes`.
+/// Then it packs these padded `bytes` and preserves the original length as the first scalar and returns the hash of the scalars via `hash_scalars`.
+///
+/// This is used when we want to preserve the length of the `bytes` to prevent collisions where `bytes` could terminate in 0's.
```

Comment:
"This is used when we want to **encode** the length of the `bytes` to prevent collisions **in cases** where `bytes` could terminate in 0's."